packages/node-core/src/indexer/testing.service.ts

```diff
@@ -0,0 +1,265 @@
+// Copyright 2020-2022 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import {readdirSync, statSync} from 'fs';
+import path from 'path';
+import {Inject, Injectable} from '@nestjs/common';
+import {
+  NodeConfig,
+  StoreService,
+  getLogger,
+  SandboxOption,
+  TestSandbox,
+  IndexerManager,
+  ISubqueryProject,
+} from '@subql/node-core';
+import {SubqlTest} from '@subql/testing/interfaces';
+import {DynamicDatasourceCreator, Store} from '@subql/types';
+import {getAllEntitiesRelations} from '@subql/utils';
+import chalk from 'chalk';
+import Pino from 'pino';
+import {CreationAttributes, Model, Sequelize} from 'sequelize';
+import {ApiService} from '../api.service';
+
+const logger = getLogger('subql-testing');
+
+declare global {
+  //const api: ApiAt;
+  const logger: Pino.Logger;
+  const store: Store;
+  const createDynamicDatasource: DynamicDatasourceCreator;
+}
+
+@Injectable()
+export abstract class TestingService<B, DS> {
+  private tests: Record<number, SubqlTest[]> = {};
+  private testSandboxes: TestSandbox[];
+  private failedTestsSummary: {
+    testName: string;
+    entityId: string;
+    entityName: string;
+    failedAttributes: string[];
+  }[] = [];
+
+  private totalTests = 0;
+  private totalPassedTests = 0;
+  private totalFailedTests = 0;
+
+  constructor(
+    protected readonly sequelize: Sequelize,
+    protected readonly nodeConfig: NodeConfig,
+    protected readonly storeService: StoreService,
+    @Inject('ISubqueryProject') protected project: ISubqueryProject<any, DS>,
+    protected readonly apiService: ApiService,
+    protected readonly indexerManager: IndexerManager<B, DS>
+  ) {}
+
+  abstract indexBlock(block: B, handler: string): Promise<void>;
+
+  async init() {
+    await this.indexerManager.start();
+    const projectPath = this.project.root;
+    // find all paths to test files
+    const testFiles = this.findAllTestFiles(path.join(projectPath, 'dist/test'));
+
+    // import all test files
+    this.testSandboxes = testFiles.map((file) => {
+      const option: SandboxOption = {
+        root: this.project.root,
+        entry: file,
+        script: null,
+      };
+
+      return new TestSandbox(option, this.nodeConfig);
+    });
+
+    logger.info(`Found ${this.testSandboxes.length} test files`);
+
+    await Promise.all(
+      this.testSandboxes.map(async (sandbox) => {
+        await sandbox.runTimeout(1000);
+      })
+    );
+
+    this.testSandboxes.map((sandbox, index) => {
+      this.tests[index] = sandbox.getTests();
+    });
+  }
```

Comment:
Could we combine this so that `runTimeout` returns the tests? Or `getTests` internally calls `runTimeout`